datafusion/core/src/physical_plan/aggregates/priority_map.rs

```diff
@@ -0,0 +1,820 @@
+                    let input_values = evaluate_many(
+                        &self.aggregate_arguments,
+                        batches.first().unwrap(),
+                    )?;
+                    let input_values = match input_values.as_slice() {
+                        [] => {
+                            Err(DataFusionError::Execution("vals required".to_string()))?
+                        }
+                        [vals] => vals,
+                        _ => {
+                            Err(DataFusionError::Execution("1 val required".to_string()))?
+                        }
+                    };
+                    let input_values = match input_values.as_slice() {
+                        [] => {
+                            Err(DataFusionError::Execution("vals required".to_string()))?
+                        }
+                        [vals] => vals,
+                        _ => {
+                            Err(DataFusionError::Execution("1 val required".to_string()))?
+                        }
+                    }
+                    .clone();
+
+                    // iterate over each column of group_by values
+                    (*self.aggregator).intern(group_by_values, input_values)?;
+                }
+                // inner is done, emit all rows and switch to producing output
+                None => {
+                    if self.aggregator.is_empty() {
+                        trace!("partition {} emit None", self.partition);
+                        return Poll::Ready(None);
+                    }
+                    let cols = self.aggregator.emit()?;
+                    let batch = RecordBatch::try_new(self.schema.clone(), cols)?;
+                    trace!(
+                        "partition {} emit batch with {} rows",
+                        self.partition,
+                        batch.num_rows()
+                    );
+                    if log::log_enabled!(Level::Trace) {
+                        print_batches(&[batch.clone()])?;
+                    }
+                    return Poll::Ready(Some(Ok(batch)));
+                }
+                // inner had error, return to caller
+                Some(Err(e)) => {
+                    return Poll::Ready(Some(Err(e)));
+                }
+            }
+        }
+        Poll::Pending
+    }
+}
+
+pub struct PriorityMap<ID: KeyType, VAL: ValueType> {
+    limit: usize,
+    sort: SortOptions,
+    rnd: RandomState,
+    id_to_val: RawTable<MapItem<ID, VAL>>,
+    val_to_idx: BTreeSet<HeapItem<VAL>>,
+}
+
+pub struct MapItem<ID: KeyType, VAL: ValueType> {
+    hash: u64,
+    pub id: ID,
+    pub val: Option<VAL>,
+}
+
+impl<ID: KeyType, VAL: ValueType> MapItem<ID, VAL> {
+    pub fn new(hash: u64, id: ID, val: Option<VAL>) -> Self {
+        Self { hash, id, val }
+    }
+}
+
+struct HeapItem<VAL: ValueType> {
+    val: Option<VAL>,
+    buk_idx: usize,
+    sort: SortOptions,
```

Comment:
I think `Sortoptions` shouldn't be stored inside `HeapItem` as it will be identical for each value

Reply:
The interface of `BTreeSet` is a bit cumbersome for us as in that you can't provide an `Ord` in a method.

Maybe the following might be a good enough compromise, which allows to do implement `cmp` without copying the `SortOptions` for each value.
```
struct HeapItem<VAL: ValueType, const DESC: bool, const NULLS_FIRST: bool> {
    val: Option<VAL>,
    buk_idx: usize,
}
```

Reply:
Downside is that we instantiate 4x the nr of types 🤔

Reply:
This is another one I am tempted to defer... if we switch away from `BTreeSet`, we won't be bound to the tyranny of the `Ord` trait. And as much as I dislike it as well, it's only two bytes 🥺 ? 

(though I don't know what the compiler is doing with alignment. I could see it being 1 byte or more likely 8)

Reply:
> , const DESC: bool, const NULLS_FIRST: bool>

A brilliant and interesting suggestion though... I didn't know you could do that!

Reply:
Actually, I really like this idea and it doesn't seem like much effort. I'll give it a shot and push shortly.

Reply:
Another option would be not changing the `Ord` at all (at least for `DESC/ASC`) but picking the worst/best value outside of the `BTreeSet` based on `SortOptions`.

Reply:
<img width="871" alt="Screenshot 2023-08-18 at 8 47 10 AM" src="https://github.com/apache/arrow-datafusion/assets/3855243/9b623872-3ef7-42c0-b0f4-d3bf99d440e3">

Oh, I see what you mean. This makes it `N^2 * 4`... perhaps it is best just to leave it on every `HeapItem` until the next PR converts the `BTreeSet` to a custom heap?

Reply:
I screwed this up: https://database.guide/max-function-in-postgresql/

`MAX()` and `MIN()` should always ignore NULL values. Problem solved!